### ISSUE TITLE
Refactoring thoughts for 5.0

### ISSUE THREAD

**Issue by ibell** [ibell/coolprop#162]
@jowr I've been starting to refactor in little ways, and I have a few other ideas:
- I want to deprecate/remove the ancillary rhosatL_anc() etc. functions, and use a class method of CoolPropStateClassSI and friends instead to get ancillaries.  We should still maintain the ability to retrieve the values from PropsSI without a saturation call though.  These are not very heavily used I don't think, except by me.
- I'd also like to deprecate/remove the PropsU and DerivU functions since we now have the SI variants
- Let's work towards deprecating Props() and all its variants and move to SI units only.  The older functions can stay around, but they will then become wrappers around the SI functions.  Oh and the default molar mass in SI is in kg/mol, which is quite weird units.  But better to stick to the standard.
- It sucks having to maintain C++ and python documentation in parallel.  It would be great if we could just write doxygen comments and then inject them into our cython code somehow.  I have some ideas along these lines
- More unit testing at the C++ level.
- Examples that test our code for as many wrappers as possible through TravisCI
- CoolPropStateClassSI should be renamed to the less verbose StateClassSI and CoolPropStateClassSI should be kept as a class inheriting from StateClassSI
- Determine what to do about the REFPROP function and the Refprop class.  The REFPROP class uses our numerical methods while the REFPROP function uses their numerical methods.  Overall I think their algorithms have been better tested, so I think it would be better to use theirs
- Generalize CoolProp, REFPROP, (and TREND?) as data engine classes derived from a base data engine class.  That way we could add another engine without too much trouble.  I'm thinking something along the lines of what is done with the solver map in ExternalMedia

**Comment by jowr** [comment]
Yes, I very much agree with these points. I suggest to divide the software into two more or less independent parts: Pure/pseudo pure fluids and mixtures. The mixture part might call the other one, but not vice versa. Furthermore, I suggest to make all the wrappers use the shared libraries. This causes some work now, but updating one shared library is so much easier than each and every wrapper.

**Comment by ibell** [comment]
What about call overhead when using shared libraries?  Passing data through
the shared library interface slows things down quite a bit, right?

**Comment by jowr** [comment]
Yes it does, but I have no feeling for how important that is. Well, let's see. If you want, you can have a look at the https://github.com/ibell/coolprop/tree/OOStructure branch. That is where I tried to sketch a possible new layout during Christmas time...

**Comment by ibell** [comment]
Yes I'll take a look at the proposed structure and do some thinking.  I
like what I see so far.

Another aggressive idea that I have very mixed feelings about is the use of
explicit units throughout the C++ codebase.  Rather than passing a double
to the T(h,p) function we pass a PressureUnit, or something like that.
There are some packages that are reasonably easy to use in C++ for units,
but I have yet to find something perfectly suited to our needs.  So we
might want to think about baking our own, or discard this idea entirely. To
be discussed further.

Another idea is to implement our fluids in JSON format rather than through
the use of hard-coded compiled values.  The rapidjson parser is really
quite nice.  So we could compile in most of the normal fluids, and then
someone could provide their own fluid(s) at runtime in JSON format.  Much
nicer than the awkward REFPROP fluid file format. I don't know what to do
about viscosity and conductivity models since they are somewhat random in
structure.  In the end we seem to be converging towards REFPROP's structure
I think.

And I made some fairly major improvements to the documentation building.
We now use IPython files to generate the fluid information.  I'll cc you on
my email to the IPython mailing list with my workflow.

**Comment by squoilin** [comment]
Don't really know how this is to be treated, but for the incompressible fluids like LiBr, it would be nice to have concentration as an input, instead of inside the filename string

**Comment by jowr** [comment]
This is done (more or less).
